src/confidential.rs

```diff
@@ -17,251 +17,231 @@
-                            x
-                        } else {
-                            return Err(A::Error::custom("missing prefix"));
-                        };
-
-                        match prefix {
-                            0 => Ok($name::Null),
-                            1 => {
-                                // Apply $explicit_fn to allow `Value` to swap the amount bytes
-                                match access.next_element()? {
-                                    Some(x) => Ok($name::Explicit($explicit_fn(x))),
-                                    None => Err(A::Error::custom("missing commitment")),
-                                }
-                            }
-                            p if p == $prefixA || p == $prefixB => {
-                                match access.next_element()? {
-                                    Some(y) => Ok($name::Confidential(p, y)),
-                                    None => Err(A::Error::custom("missing commitment")),
-                                }
-                            }
-                            p => return Err(A::Error::custom(format!(
-                                "invalid commitment, invalid prefix: 0x{:02x}", p
-                            ))),
-                        }
-                    }
-                }
+            buffer[0]
+        };
 
-                d.deserialize_seq(CommitVisitor)
+        match prefix {
+            0 => {
+                // consume null value prefix
+                d.consume(1);
+                Ok(Value::Null)
+            }
+            1 => {
+                // ignore prefix when decoding an explicit value
+                d.consume(1);
+                let explicit = u64::swap_bytes(Decodable::consensus_decode(&mut d)?);
+                Ok(Value::Explicit(explicit))
+            }
+            p if p == 0x08 || p == 0x09 => {
+                let bytes = <[u8; 33]>::consensus_decode(&mut d)?;
+                Ok(Value::Confidential(PedersenCommitment::from_slice(&bytes)?))
             }
+            p => return Err(encode::Error::InvalidConfidentialPrefix(p)),
         }
-    );
+    }
 }
 
-/// A CT commitment to an amount
-#[derive(Copy, Clone, Debug, Eq, Hash, PartialEq, PartialOrd, Ord)]
-pub enum Value {
-    /// No value
-    Null,
-    /// Value is explicitly encoded
-    Explicit(u64),
-    // Split commitments into a 1-byte prefix and 32-byte commitment, because
-    // they're easy enough to separate and Rust stdlib treats 32-byte arrays
-    // much much better than 33-byte arrays.
-    /// Value is committed
-    Confidential(u8, [u8; 32]),
-}
-impl_confidential_commitment!(Value, u64, 0x08, 0x09, u64::swap_bytes);
+#[cfg(feature = "serde")]
+impl Serialize for Value {
+    fn serialize<S: Serializer>(&self, s: S) -> Result<S::Ok, S::Error> {
+        use serde::ser::SerializeSeq;
+
+        let seq_len = if *self == Value::Null { 1 } else { 2 };
```

Comment:
Thanks for this comment! This has highlighted that we had introduced some bugs in the serde code which were not showing up because of some missing tests. I'm adding these tests to the split up, less controversial PR which I'll open soon.